app/service/rest.py

```diff
@@ -270,3 +280,39 @@ def get_detailed_services():
             service.statistics = statistics.create_zeroed_stats_dicts()
 
     return detailed_service_schema.dump(services.values(), many=True).data
+
+
+@service_blueprint.route('/<uuid:service_id>/whitelist', methods=['GET'])
+def get_whitelist(service_id):
+    service = dao_fetch_service_by_id(service_id)
+
+    if not service:
+        raise InvalidRequest("Service does not exist", status_code=404)
+
+    whitelist = dao_fetch_service_whitelist(service.id)
+    return jsonify(
+        email_addresses=[item.recipient for item in whitelist
+                         if item.recipient_type == EMAIL_TYPE],
+        phone_numbers=[item.recipient for item in whitelist
+                       if item.recipient_type == MOBILE_TYPE]
+    )
+
+
+@service_blueprint.route('/<uuid:service_id>/whitelist', methods=['PUT'])
+def update_whitelist(service_id):
+    # doesn't commit so if there are any errors, we preserve old values in db
+    dao_remove_service_whitelist(service_id)
+    try:
+        whitelist_objs = itertools.chain(
+            [ServiceWhitelist.from_string(service_id, MOBILE_TYPE, recipient)
+             for recipient in request.get_json().get('phone_numbers')],
+            [ServiceWhitelist.from_string(service_id, EMAIL_TYPE, recipient)
+             for recipient in request.get_json().get('email_addresses')])
```

Comment:
I think this would be easier to read if you factored out some of the code into helper functions, eg:

```python
def get_recipients_from_request(request_json, key, type):
  return [(type, recipient) for recipient in request_json.get(key)]

def get_whitelist_objects(service_id, request_json):
  return [
    ServiceWhitelist.from_string(service_id, type, recipient)
    for type, recipient in (
      get_recipients_from_request(request_json, 'phone_numbers', MOBILE_TYPE) +
      get_recipients_from_request(request_json, 'email_addresses', EMAIL_TYPE)
    )
  ]

whitelist_objs = get_whitelist_objects(service_id, request.get_json())
```